staking/staking-specification.md

```diff
@@ -0,0 +1,815 @@
+
+```
+/// @dev Set all configurable parameters at once.
+/// @param _epochDurationInSeconds Minimum seconds between epochs.
+/// @param _rewardDelegatedStakeWeight How much delegated stake is weighted vs operator stake, in ppm.
+/// @param _minimumPoolStake Minimum amount of stake required in a pool to collect rewards.
+/// @param _maximumMakersInPool Maximum number of maker addresses allowed to be registered to a pool.
+/// @param _cobbDouglasAlphaNumerator Numerator for cobb douglas alpha factor.
+/// @param _cobbDouglasAlphaDenominator Denominator for cobb douglas alpha factor.
+function setParams(
+    uint256 _epochDurationInSeconds,
+    uint32 _rewardDelegatedStakeWeight,
+    uint256 _minimumPoolStake,
+    uint256 _maximumMakersInPool,
+    uint32 _cobbDouglasAlphaNumerator,
+    uint32 _cobbDouglasAlphaDenominator
+) external;
+
+/// @dev Retrieves all configurable parameter values.
+/// @return _epochDurationInSeconds Minimum seconds between epochs.
+/// @return _rewardDelegatedStakeWeight How much delegated stake is weighted vs operator stake, in ppm.
+/// @return _minimumPoolStake Minimum amount of stake required in a pool to collect rewards.
+/// @return _maximumMakersInPool Maximum number of maker addresses allowed to be registered to a pool.
+/// @return _cobbDouglasAlphaNumerator Numerator for cobb douglas alpha factor.
+/// @return _cobbDouglasAlphaDenominator Denominator for cobb douglas alpha factor.
+function getParams()
+    external
+    view
+    returns (
+        uint256 _epochDurationInSeconds,
+        uint32 _rewardDelegatedStakeWeight,
+        uint256 _minimumPoolStake,
+        uint256 _maximumMakersInPool,
+        uint32 _cobbDouglasAlphaNumerator,
+        uint32 _cobbDouglasAlphaDenominator
+    );
+```
+
+
+## 4 Epochs & Scheduling
+
+All processes in the system are segmented into nonoverlapping time intervals, called epochs. Epochs have a fixed minimum period (10 days at time of writing), which is configurable via [MixinParams](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/sys/MixinParams.sol). Epochs serve as the basis for all other timeframes within the system, which provides a more stable and consistent scheduling metric than timestamps.
+
+<p align="center"><img src="./assets/Epochs.png" width="700" /></p>
+
+### 4.1 Ending One Epoch, and Starting a New One
+
+A new epoch automatically begins when the current epoch ends. Anyone can "end" an epoch by calling the staking contract after the minimum epoch period has elapsed.
+
+```
+function endEpoch()
+    external
+    returns (uint256 poolsRemaining);
+```
+
+Note: The return value (`poolsRemaining`) is described in detail in Section 7 Liquidity Rewards.
+
+
+## 5 Staking
+
+ZRX is staked by depositing tokens into the Staking Contract. The diagram below illustrates this workflow.
+
+<p align="center"><img src="./assets/Staking.png" width="700" /></p>
```

Comment:
The diagram is really nice, but it would be nice to differentiate the `Staker Balance` in the `ZrxVault` and the `Staker Balance` in the `Zrx Token`. I think that readers could get confused by the difference between these balances.